datafusion/core/src/execution/context.rs

```diff
@@ -1466,14 +1466,15 @@ impl SessionState {
         }
 
         let mut rules: Vec<Arc<dyn OptimizerRule + Sync + Send>> = vec![
-            // Simplify expressions first to maximize the chance
-            // of applying other optimizations
-            Arc::new(SimplifyExpressions::new()),
-            Arc::new(PreCastLitInComparisonExpressions::new()),
+            // TODO https://github.com/apache/arrow-datafusion/issues/3557#issuecomment-1259227250
+            // type coercion can't handle the subquery plan, should rewrite subquery first
```

Comment:
I want to move the `type coercion` to the beginning of all rules, but we can't handle the subquery for it.
So need to do some subquery rule first as the intermediate state.

cc @andygrove  Do you have any suggestion for that.

Reply:
I know less about the subquery.